include/ginkgo/core/config/property_tree.hpp

```diff
@@ -0,0 +1,210 @@
+
+
+#include <cstdint>
+#include <limits>
+#include <map>
+#include <stdexcept>
+#include <string>
+#include <type_traits>
+#include <vector>
+
+
+namespace gko {
+namespace config {
+
+
+/**
+ * pnode describes a tree of properties.
+ *
+ * A pnode can either be empty, hold a value (a string, integer, real, or bool),
+ * contain an array of pnode., or contain a mapping between strings and pnodes.
+ */
+class pnode {
+public:
+    using key_type = std::string;
+    using map_type = std::map<key_type, pnode>;
+    using array_type = std::vector<pnode>;
+
+    /**
+     * status_t is the indicator for the current node storage.
+     */
+    enum class status_t { empty, array, boolean, real, integer, string, map };
+
+    /**
+     * Default constructor: create an empty node
+     */
+    pnode();
+
+    /**
+     * Constructor for bool
+     *
+     * @param boolean  the bool type value
+     */
+    pnode(bool boolean);
+
+    /**
+     * Constructor for integer with all integer type
+     *
+     * @tparam T  input type
+     *
+     * @param integer  the integer type value
+     */
+    template <typename T,
+              std::enable_if_t<std::is_integral<T>::value>* = nullptr>
+    pnode(T integer);
+
+    /**
+     * Constructor for string
+     *
+     * @param str  string type value
+     */
+    pnode(const std::string& str);
+
+    /**
+     * Constructor for char* (otherwise, it will use bool)
+     *
+     * @param str  the string like "..."
+     */
+    pnode(const char* str);
+
+    /**
+     * Constructor for double (and also float)
+     *
+     * @param real  the floating point type value
+     */
+    pnode(double real);
+
+    /**
+     * Constructor for array
+     *
+     * @param array  an pnode array
+     */
+    pnode(const array_type& array);
+
+    /**
+     * Constructor for map
+     *
+     * @param map  a (string, pnode)-map
+     */
+    pnode(const map_type& map);
+
+    /**
+     * bool conversion. It's true if and only if it is not empty.
+     */
+    operator bool() const noexcept;
+
+    /**
+     * Get the current node status.
+     *
+     * @return the status
+     */
+    status_t get_status() const;
+
+    /**
+     * Get the array. It will throw error if the current node does not hold an
+     * array.
+     *
+     * @return the array const reference
+     */
+    const array_type& get_array() const;
+
+    /**
+     * Get the map. It will throw error if the current node does not hold an
+     * map.
+     *
+     * @return the map const reference
+     */
+    const map_type& get_map() const;
+
+    /**
+     * Get the boolean value. It will throw error if the current node does not
+     * hold an boolean value.
+     *
+     * @return the boolean value
+     */
+    bool get_boolean() const;
+
+    /**
+     * Get the integer value. It will throw error if the current node does not
+     * hold an integer value.
+     *
+     * @return the integer value with type int64_t
```

Comment:
we can see the return type from the function signature, which should be visible everywhere it matters (IDEs, Doxygen)
```suggestion
     * @return the integer value
```